pageserver/src/virtual_file/owned_buffers_io/write.rs

```diff
@@ -53,61 +54,121 @@ where
         Ok(writer)
     }
 
-    pub async fn write_buffered<B: IoBuf>(&mut self, chunk: Slice<B>) -> std::io::Result<()>
+    #[inline(always)]
+    fn buf(&self) -> &B {
+        self.buf
+            .as_ref()
+            .expect("must not use after we returned an error")
+    }
+
+    pub async fn write_buffered<S: IoBuf>(&mut self, chunk: Slice<S>) -> std::io::Result<(usize, S)>
```

Comment:
nit: Don't mean to open a can of worms, but how is the caller supposed to handle failures? Since `Slice` takes ownership of the buffer at the call site, an error essentially means the caller lost the buffer. It follows that callers should keep a copy if they want to handle errors. Am I missing something?

Reply:
Yeah, it's a wrinkle, but a pre-existing one.

The (only) correct way to handle errors is to bail.
Since
1. we're only using this to write to local disk and
2. we're aborting the process on local disk IO errors anyway,

I'd say let's worry about it later.

(The only user as of this PR is the on-demand download path, which on either download or disk error bails out, deletes the partial download, and retries after a back-off)

Reply:
> we're aborting the process on local disk IO errors anyway

Yeah. I thought that would be our saving grace. Let's punt on this for now, although a comment or a follow up issue would be nice.